contracts/schemes/JoinAndQuit.sol

```diff
@@ -66,51 +67,73 @@ contract JoinAndQuit is
-     * @param _votingMachine the voting machines address to
-     * @param _votingParams genesisProtocol parameters - valid only if _voteParamsHash is zero
-     * @param _voteOnBehalf genesisProtocol parameter - valid only if _voteParamsHash is zero
-     * @param _voteParamsHash voting machine parameters.
-     * @param _fundingToken the funding token - if this is zero the donation will be in native token ETH
-     * @param _minFeeToJoin minimum fee required to join
-     * @param _memberReputation the repution which will be allocated for members
-              if this param is zero so the repution will be allocated proportional to the fee paid
-     * @param _fundingGoal the funding goal
-     * @param _fundingGoalDeadline the funding goal deadline
-     * @param _rageQuitEnable rageQuit enabling flag
+     * @param _votingParams genesisProtocol parameters
+     * @param _voteOnBehalf  parameter
+     * @param _daoFactory  DAOFactory instance to instance a votingMachine.
+     * @param _stakingToken (for GenesisProtocol)
+     * @param _packageVersion packageVersion to instance the votingMachine from.
+     * @param _votingMachineName the votingMachine contract name.
+     * @param _encodedJoinAndQuitParams encoded JoinAndQuitParams
      */
     function initialize(
         Avatar _avatar,
-        IntVoteInterface _votingMachine,
         uint256[11] calldata _votingParams,
         address _voteOnBehalf,
-        bytes32 _voteParamsHash,
-        IERC20 _fundingToken,
-        uint256 _minFeeToJoin,
-        uint256 _memberReputation,
-        uint256 _fundingGoal,
-        uint256 _fundingGoalDeadline,
-        bool    _rageQuitEnable
+        DAOFactory _daoFactory,
+        address _stakingToken,
+        uint64[3] calldata _packageVersion,
+        string calldata _votingMachineName,
+        bytes calldata _encodedJoinAndQuitParams
     )
     external
     {
-        super._initializeGovernance(_avatar, _votingMachine, _voteParamsHash, _votingParams, _voteOnBehalf);
-        fundingToken = _fundingToken;
-        minFeeToJoin = _minFeeToJoin;
-        memberReputation = _memberReputation;
-        fundingGoal = _fundingGoal;
-        fundingGoalDeadline = _fundingGoalDeadline;
-        rageQuitEnable = _rageQuitEnable;
+        super._initializeGovernance(
+            _avatar,
+            _votingParams,
+            _voteOnBehalf,
+            _daoFactory,
+            _stakingToken,
+            address(this),
+            address(this),
+            address(this),
+            _packageVersion,
+            _votingMachineName);
+
+        (
+            address  fundingToken,
+            uint256  minFeeToJoin,
+            uint256  memberReputation,
+            uint256  fundingGoal,
+            uint256  fundingGoalDeadline,
+            bool  rageQuitEnable
+        )  =
+        abi.decode(_encodedJoinAndQuitParams, (address, uint256, uint256, uint256, uint256, bool));
```

Comment:
Why is that? Not that I see an issue, just want to understand why the added complexity here?

Reply:
this is needed to solve stack too deep issue.